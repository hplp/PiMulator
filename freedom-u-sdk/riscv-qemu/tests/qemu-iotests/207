#!/bin/bash
#
# Test ssh image creation
#
# Copyright (C) 2018 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

# creator
owner=kwolf@redhat.com

seq=`basename $0`
echo "QA output created by $seq"

here=`pwd`
status=1	# failure is the default!

# get standard environment, filters and checks
. ./common.rc
. ./common.filter

_supported_fmt raw
_supported_proto ssh
_supported_os Linux

function do_run_qemu()
{
    echo Testing: "$@"
    $QEMU -nographic -qmp stdio -serial none "$@"
    echo
}

function run_qemu()
{
    do_run_qemu "$@" 2>&1 | _filter_testdir | _filter_qmp \
                          | _filter_qemu | _filter_imgfmt \
                          | _filter_actual_image_size
}

echo
echo "=== Successful image creation (defaults) ==="
echo

run_qemu <<EOF
{ "execute": "qmp_capabilities" }
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "$TEST_IMG_FILE",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          }
      },
      "size": 4194304
  }
}
{ "execute": "quit" }
EOF

_img_info | _filter_img_info
echo
TEST_IMG=$TEST_IMG_FILE _img_info | _filter_img_info

echo
echo "=== Test host-key-check options ==="
echo

run_qemu <<EOF
{ "execute": "qmp_capabilities" }
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "$TEST_IMG_FILE",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          },
          "host-key-check": {
              "mode": "none"
          }
      },
      "size": 8388608
  }
}
{ "execute": "quit" }
EOF

_img_info | _filter_img_info

run_qemu <<EOF
{ "execute": "qmp_capabilities" }
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "$TEST_IMG_FILE",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          },
          "host-key-check": {
              "mode": "known_hosts"
          }
      },
      "size": 4194304
  }
}
{ "execute": "quit" }
EOF

_img_info | _filter_img_info


key=$(ssh-keyscan -t rsa 127.0.0.1 2>/dev/null | grep -v "\\^#" |
      cut -d" " -f3 | base64 -d | md5sum -b | cut -d" " -f1)

run_qemu <<EOF
{ "execute": "qmp_capabilities" }
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "$TEST_IMG_FILE",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          },
          "host-key-check": {
              "mode": "hash",
              "type": "md5",
              "hash": "wrong"
          }
      },
      "size": 8388608
  }
}
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "$TEST_IMG_FILE",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          },
          "host-key-check": {
              "mode": "hash",
              "type": "md5",
              "hash": "$key"
          }
      },
      "size": 8388608
  }
}
{ "execute": "quit" }
EOF

_img_info | _filter_img_info


key=$(ssh-keyscan -t rsa 127.0.0.1 2>/dev/null | grep -v "\\^#" |
      cut -d" " -f3 | base64 -d | sha1sum -b | cut -d" " -f1)

run_qemu <<EOF
{ "execute": "qmp_capabilities" }
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "$TEST_IMG_FILE",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          },
          "host-key-check": {
              "mode": "hash",
              "type": "sha1",
              "hash": "wrong"
          }
      },
      "size": 4194304
  }
}
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "$TEST_IMG_FILE",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          },
          "host-key-check": {
              "mode": "hash",
              "type": "sha1",
              "hash": "$key"
          }
      },
      "size": 4194304
  }
}
{ "execute": "quit" }
EOF

_img_info | _filter_img_info

echo
echo "=== Invalid path and user ==="
echo

run_qemu <<EOF
{ "execute": "qmp_capabilities" }
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "/this/is/not/an/existing/path",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          }
      },
      "size": 4194304
  }
}
{ "execute": "x-blockdev-create",
  "arguments": {
      "driver": "ssh",
      "location": {
          "path": "$TEST_IMG_FILE",
          "user": "invalid user",
          "server": {
              "host": "127.0.0.1",
              "port": "22"
          }
      },
      "size": 4194304
  }
}
{ "execute": "quit" }
EOF

# success, all done
echo "*** done"
rm -f $seq.full
status=0
