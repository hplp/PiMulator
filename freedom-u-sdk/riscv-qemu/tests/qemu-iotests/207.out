QA output created by 207

=== Successful image creation (defaults) ===

Testing:
QMP_VERSION
{"return": {}}
{"return": {}}
{"return": {}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "SHUTDOWN", "data": {"guest": false}}

image: json:{"driver": "IMGFMT", "file": {"server.host": "127.0.0.1", "server.port": "22", "driver": "ssh", "path": "TEST_DIR/t.IMGFMT"}}
file format: IMGFMT
virtual size: 4.0M (4194304 bytes)

image: TEST_DIR/t.IMGFMT
file format: IMGFMT
virtual size: 4.0M (4194304 bytes)

=== Test host-key-check options ===

Testing:
QMP_VERSION
{"return": {}}
{"return": {}}
{"return": {}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "SHUTDOWN", "data": {"guest": false}}

image: json:{"driver": "IMGFMT", "file": {"server.host": "127.0.0.1", "server.port": "22", "driver": "ssh", "path": "TEST_DIR/t.IMGFMT"}}
file format: IMGFMT
virtual size: 8.0M (8388608 bytes)
Testing:
QMP_VERSION
{"return": {}}
{"return": {}}
{"return": {}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "SHUTDOWN", "data": {"guest": false}}

image: json:{"driver": "IMGFMT", "file": {"server.host": "127.0.0.1", "server.port": "22", "driver": "ssh", "path": "TEST_DIR/t.IMGFMT"}}
file format: IMGFMT
virtual size: 4.0M (4194304 bytes)
Testing:
QMP_VERSION
{"return": {}}
{"error": {"class": "GenericError", "desc": "remote host key does not match host_key_check 'wrong'"}}
{"return": {}}
{"return": {}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "SHUTDOWN", "data": {"guest": false}}

image: json:{"driver": "IMGFMT", "file": {"server.host": "127.0.0.1", "server.port": "22", "driver": "ssh", "path": "TEST_DIR/t.IMGFMT"}}
file format: IMGFMT
virtual size: 8.0M (8388608 bytes)
Testing:
QMP_VERSION
{"return": {}}
{"error": {"class": "GenericError", "desc": "remote host key does not match host_key_check 'wrong'"}}
{"return": {}}
{"return": {}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "SHUTDOWN", "data": {"guest": false}}

image: json:{"driver": "IMGFMT", "file": {"server.host": "127.0.0.1", "server.port": "22", "driver": "ssh", "path": "TEST_DIR/t.IMGFMT"}}
file format: IMGFMT
virtual size: 4.0M (4194304 bytes)

=== Invalid path and user ===

Testing:
QMP_VERSION
{"return": {}}
{"error": {"class": "GenericError", "desc": "failed to open remote file '/this/is/not/an/existing/path': Failed opening remote file (libssh2 error code: -31)"}}
{"error": {"class": "GenericError", "desc": "failed to authenticate using publickey authentication and the identities held by your ssh-agent"}}
{"return": {}}
{"timestamp": {"seconds":  TIMESTAMP, "microseconds":  TIMESTAMP}, "event": "SHUTDOWN", "data": {"guest": false}}

*** done
